### PR TITLE
Distributor: Loki API can receive gzipped JSON

### LINKED ISSUE #3205: Please add compression ability in push API
**Is your feature request related to a problem? Please describe.**
I need send very big logs.
When I send it to Loki, I have high network traffic between servers.

**Describe the solution you'd like**
I want send compressed POST body (JSON in my case) to API for reduce network traffic.
I try GZIP logs. File reduced to 25-30 times from origin size.

Gzipped API can have some extension: /loki/api/v1/push.gz
Or use HTTP request header "Content-Encoding: gzip"

**Describe alternatives you've considered**
It may be proprietary HTTP or TCP service on same node (where runs Loki), which will receive GZIPped log data (in custom format), decompress it and send to Loki.

**Additional context**

### PR DESCRIPTION
**Pull request by ukolovda**
**What this PR does / why we need it**:

This PR allow send gzipped `JSON` to push endpoint.

HTTP server recognize `Content-Encoding: gzip` request header (for JSON only) and unpack request body.
This option can reduce network traffic (5-30 times) and can be useful for sending logs to remote server.

**Which issue(s) this PR fixes**:

Fixes #3205

**Special notes for your reviewer**:

**Checklist**
- [x] Documentation added
- [ ] Tests updated

### CODE REVIEW

File: pkg/distributor/http.go [hunk 1]
```diff
@@ -97,6 +117,7 @@ func ParseRequest(r *http.Request) (*logproto.PushRequest, error) {
 			"msg", "push request parsed",
 			"path", r.URL.Path,
 			"contentType", contentType,
+			"contentEncoding", contentEncoding,
```
Comment (cyriltovena) [inline pkg/distributor/http.go:120]: ❤️

**Review by owen-d — APPROVED** [review]
This looks great, thanks!

I notice we don't have an `http_test.go` file, so I won't ask that you include some additional testing for this unless you'd like to.

Follow ups for @cyriltovena & myself:
- use gzip pools
- test gzip endcoded proto & json bodies

**Comment by ukolovda** [comment]
> This looks great, thanks!
> 
> I notice we don't have an `http_test.go` file, so I won't ask that you include some additional testing for this unless you'd like to.
> 
> Follow ups for @CyrilPeponnet & myself:
> 
>     * use gzip pools
> 
>     * test gzip endcoded proto & json bodies

Thank you!
I'm beginning in Go and didn't make tests for HTTP server yet...
I made manual tests, then deploy patched version to my server. It works fine.

**Review by cyriltovena — APPROVED** [review]
LGTM

@owen-d anything you wanted to comment on ? if not please merge.

**Review by cyriltovena — COMMENTED** [review]
LGTM

**Pull request merged**